transit/world_feed/gtfs_converter/gtfs_converter.cpp

```diff
@@ -0,0 +1,262 @@
+  {
+    LOG(LINFO, ("Could not parse stops.", res.message));
+    return FeedStatus::CORRUPTED;
+  }
+
+  if (auto res = feed.read_stop_times(); res != gtfs::ResultCode::OK)
+  {
+    LOG(LINFO, ("Could not parse stop times.", res.message));
+    return FeedStatus::CORRUPTED;
+  }
+
+  // We try to parse optional for json files and do not return error in case of invalid file
+  // content, only log warning message.
+  if (auto res = feed.read_calendar(); gtfs::ErrorParsingOptionalFile(res))
+    LOG(LINFO, ("Could not parse calendar.", res.message));
+
+  if (auto res = feed.read_calendar_dates(); gtfs::ErrorParsingOptionalFile(res))
+    LOG(LINFO, ("Could not parse calendar dates.", res.message));
+
+  if (auto res = feed.read_frequencies(); gtfs::ErrorParsingOptionalFile(res))
+    LOG(LINFO, ("Could not parse frequencies.", res.message));
+
+  if (auto res = feed.read_transfers(); gtfs::ErrorParsingOptionalFile(res))
+    LOG(LINFO, ("Could not parse transfers.", res.message));
+
+  if (feed.read_feed_info() == gtfs::ResultCode::OK)
+    LOG(LINFO, ("Feed info is present."));
+
+  return FeedStatus::OK;
+}
+
+int main(int argc, char ** argv)
+{
+  google::SetUsageMessage("Reads GTFS feeds, produces json with global ids.");
+  google::ParseCommandLineFlags(&argc, &argv, true);
+
+  if (FLAGS_path_mapping.empty() || FLAGS_path_gtfs_feeds.empty() || FLAGS_path_json.empty())
+  {
+    LOG(LERROR, ("Required options are not present."));
+    return -1;
+  }
+
+  if (!Platform::IsDirectory(FLAGS_path_gtfs_feeds) || !Platform::IsDirectory(FLAGS_path_json) ||
+      !Platform::IsDirectory(FLAGS_path_resources))
+  {
+    LOG(LERROR,
+        ("Check directories", FLAGS_path_gtfs_feeds, FLAGS_path_json, FLAGS_path_resources));
+    return -1;
+  }
+
+  auto const gtfsFeeds = GetGtfsFeedsInDirectory(FLAGS_path_gtfs_feeds);
+
+  if (gtfsFeeds.empty())
+  {
+    LOG(LERROR, ("No subdirectories with GTFS feeds found in", FLAGS_path_gtfs_feeds));
+    return -1;
+  }
+
+  std::vector<std::string> invalidFeeds;
+
+  size_t feedsWithNoShapesCount = 0;
+  size_t feedsNotDumpedCount = 0;
+  size_t feedsDumped = 0;
+  size_t feedsTotal = gtfsFeeds.size();
+  bool pass = true;
+
+  transit::IdGenerator generator(FLAGS_path_mapping);
+
+  GetPlatform().SetResourceDir(FLAGS_path_resources);
+  transit::ColorPicker colorPicker;
+
+  for (size_t i = 0; i < gtfsFeeds.size(); ++i)
+  {
+    base::Timer feedTimer;
+    auto feedPath = gtfsFeeds[i];
+
+    if (SkipFeed(feedPath, pass))
+    {
+      feedsTotal--;
+      LOG(LINFO, ("Skipped", feedPath));
+      continue;
+    }
+
+    bool stop = StopOnFeed(feedPath);
+    if (stop)
+      feedsTotal -= (gtfsFeeds.size() - i - 1);
+
+    ExtendPath(feedPath);
+    LOG(LINFO, ("Handling feed", feedPath));
+
+    gtfs::Feed feed(feedPath);
+
+    if (auto const res = ReadFeed(feed); res != FeedStatus::OK)
+    {
+      if (res == FeedStatus::NO_SHAPES)
+        feedsWithNoShapesCount++;
```

Comment:
++ feedsWithNoShapesCount?

Reply:
И ниже?

Reply:
Это же оптимизирует компилятор.

Reply:
Всегда? Это может оптимизировать компилятор?

Reply:
И в чем выгода у `i++`?

Reply:
Да, если i - простой тип, то компилятор гарантированно заоптимизирует. Слышала это на какой-то конфе, на SO есть похожее заключение (не все то правда что пишут на SO, но все же):
https://stackoverflow.com/questions/24901/is-there-a-performance-difference-between-i-and-i-in-c

Reply:
У варианта i++ плюсов нет, я просто не задумываюсь, когда пишу) Поменяю на ++i, хотя у нас в код стайле про простые типы и инкремент кажется ничего нет.

Reply:
Короче я к тому, что если по таким мелочам ревьюить эти постепенно растущие 3к строк, то процесс будет расходящимся))

Reply:
> Да, если i - простой тип, то компилятор гарантированно заоптимизирует.

Ага. Спасибо! Надо бы почитать как-нибудь, какие гарантии дает стандарт на эту тему.

С другой стороны у нас google code style, там есть про это:
https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement

`When the return value is ignored, the "pre" form (++i) is never less efficient than the "post" form (i++), and is often more efficient. This is because post-increment (or decrement) requires a copy of i to be made, which is the value of the expression. If i is an iterator or other non-scalar type, copying i could be expensive. Since the two types of increment behave the same when the value is ignored, why not just always pre-increment?`

`For simple scalar (non-object) values there is no reason to prefer one form and we allow either. For iterators and other template types, use pre-increment.`

От себя могу добавить, что как мне кажется, на сегодня (1) в сообществе скорее вызовут вопросы пост-инкремент, чем пре-инкремент. А вопросы равно обсуждения, PR исправляющие это и т.п. (2) В нашем проекте при прочих равных используется пре-инкремент; (3) Используя пре-инкремент мы всегда не прогадаем. Т.е. будет не хуже для простых типов. А вот для итераторов - часто лучше.

Вообще ревью PR из 3000 строк процесс обычно не быстрый. Да и хорошо они проревьены не будут, к сожалению.